repair/repair.cc

```diff
@@ -1989,26 +2024,81 @@ future<> repair_service::do_rebuild_replace_with_repair(locator::token_metadata_
+                            if (source_dc.force()) {
+                                rlogger.warn("{}: using source_dc anyway according to the force option", msg);
+                            } else if (source_dc.user_provided()) {
+                                throw std::runtime_error(format("{}: Omit the source_dc option to use all nodes in the cluster, or use the --force option to enforce it", msg));
+                            } else {
+                                rlogger.warn("{}: falling back to repair using all nodes", msg);
+                                source_dc_for_keyspace = "";
+                            }
+                        }
+                    }
+                    break;
+                }
+                case locator::replication_strategy_type::everywhere_topology:
+                    // source_dc_live_nodes is not empty, so we can use any remaining nodes
+                    source_dc_for_keyspace = *source_dc;
+                    break;
+                case locator::replication_strategy_type::simple:
+                    // With simple strategy, we have no assurance that source_dc will contain
+                    // another replica for all token ranges.
+                    break;
+                default:
+                    break;
+                }
+            }
+            rlogger.info("{}: started with keyspace={}, source_dc={}, alt_dc={}, nr_ranges={}, ignore_nodes={}", op, keyspace_name, source_dc_for_keyspace, alt_dc_for_keyspace, ranges.size() * nr_tables, ignore_nodes);
             for (auto it = ranges.begin(); it != ranges.end();) {
                 auto& r = *it;
                 seastar::thread::maybe_yield();
                 auto end_token = r.end() ? r.end()->value() : dht::maximum_token();
-                auto neighbors = boost::copy_range<std::vector<gms::inet_address>>(strat.calculate_natural_ips(end_token, *tmptr).get() |
-                    boost::adaptors::filtered([myip, &source_dc, &topology, &ignore_nodes] (const gms::inet_address& node) {
+                auto natural_ips = strat.calculate_natural_ips(end_token, *tmptr).get();
+                auto neighbors = boost::copy_range<std::vector<gms::inet_address>>(natural_ips |
+                    boost::adaptors::filtered([&] (const gms::inet_address& node) {
                         if (node == myip) {
                             return false;
                         }
                         if (ignore_nodes.contains(node)) {
                             return false;
                         }
-                        return source_dc.empty() ? true : topology.get_datacenter(node) == source_dc;
+                        return source_dc_for_keyspace.empty() || live_nodes_per_dc[source_dc_for_keyspace].contains(node)
+                                || (!alt_dc_for_keyspace.empty() && live_nodes_per_dc[alt_dc_for_keyspace].contains(node));
```

Comment:
For rebuild, when alt_dc_for_keyspace is set, the source_dc_for_keyspace is also set. This means we will build from both dcs. Is this what you meant to do? If the alt_dc_for_keyspace is better (with no node down), I think it is enough to sync data only from the alt_dc_for_keyspace.

Reply:
> For rebuild, when alt_dc_for_keyspace is set, the source_dc_for_keyspace is also set. This means we will build from both dcs. Is this what you meant to do?

Yes.  The idea is that alt_dc_for_keyspace is considered only when source_dc_for_keyspace is non-empty as well.

> If the alt_dc_for_keyspace is better (with no node down), I think it is enough to sync data only from the alt_dc_for_keyspace.

That's indeed possible.
In this case, I think we wouldn't need another variable, and more complex logic.
We can simply set source_dc_for_keyspace to the alternative dc.

Reply:
done in v8. alt_dc replaces source_dc rather than extending it